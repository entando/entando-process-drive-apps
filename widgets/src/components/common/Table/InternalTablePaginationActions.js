import PropTypes from 'prop-types';
import React from 'react';
import PlayArrowIcon from '@material-ui/icons/PlayArrow';
import SkipNextIcon from '@material-ui/icons/SkipNext';
import SkipPreviousIcon from '@material-ui/icons/SkipPrevious';
import withStyles from '@material-ui/core/styles/withStyles';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import i18next from 'i18next';

const styles = {
  root: {
    marginLeft: 'auto',
    marginRight: 20,
    display: 'flex',
    alignItems: 'center',
  },
  previousIcon: {
    transform: 'rotate(180deg)',
  },
  pageNumberInput: {
    marginRight: 10,
  },
};

class InternalTablePaginationActions extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      pageNumber: props && props.page + 1,
    };
  }

  static getDerivedStateFromProps(props, state) {
    const lastPage = Math.floor(props.count / props.rowsPerPage);
    if (props.page === lastPage || props.page !== +state.pageNumber + 1) {
      return { pageNumber: props.page + 1 };
    }
    return state;
  }

  componentDidMount() {
    const { page } = this.props;
    this.setPageNumber(page + 1);
  }

  get numberOfPages() {
    const { count, rowsPerPage } = this.props;
    return Math.ceil(count / rowsPerPage);
  }

  setPageNumber = pageNumber => {
    this.setState({ pageNumber });
  };

  handlePageNumberInputChange = event => {
    const { value } = event.target;
    const { onChangePage } = this.props;
    // sync state and entered value
    this.setPageNumber(value);
    // navigate to entered page number if valid
    const pageNumber = +value;
    if (pageNumber && pageNumber <= this.numberOfPages) {
      onChangePage(pageNumber - 1);
    }
  };

  handleFirstPageButtonClick = () => {
    const { onChangePage } = this.props;
    onChangePage(0);
  };

  handlePrevButtonClick = () => {
    const { onChangePage, page } = this.props;
    onChangePage(page - 1);
  };

  handleNextButtonClick = () => {
    const { onChangePage, page } = this.props;
    onChangePage(page + 1);
  };

  handleLastPageButtonClick = () => {
    const { onChangePage, count, rowsPerPage } = this.props;
    onChangePage(Math.max(0, Math.ceil(count / rowsPerPage) - 1));
  };

  render() {
    const { page, count, rowsPerPage, classes } = this.props;
    const { pageNumber } = this.state;
    const numberOfDigits = `${this.numberOfPages}`.length;

    return (
      <div className={classes.root}>
        <IconButton size="small" onClick={this.handleFirstPageButtonClick} disabled={page === 0}>
          <SkipPreviousIcon />
        </IconButton>

        <IconButton size="small" onClick={this.handlePrevButtonClick} disabled={page === 0}>
          <PlayArrowIcon className={classes.previousIcon} />
        </IconButton>

        <TextField
          className={classes.pageNumberInput}
          style={{ width: 25 + 10 * numberOfDigits }}
          value={pageNumber}
          onChange={event => this.handlePageNumberInputChange(event)}
        />

        <Typography>{`${i18next.t('common.of')} ${this.numberOfPages}`}</Typography>

        <IconButton
          size="small"
          onClick={this.handleNextButtonClick}
          disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        >
          <PlayArrowIcon />
        </IconButton>

        <IconButton
          size="small"
          onClick={this.handleLastPageButtonClick}
          disabled={page >= Math.ceil(count / rowsPerPage) - 1}
        >
          <SkipNextIcon />
        </IconButton>
      </div>
    );
  }
}

InternalTablePaginationActions.propTypes = {
  classes: PropTypes.shape({
    root: PropTypes.string,
    previousIcon: PropTypes.string,
    pageNumberInput: PropTypes.string,
  }),
  count: PropTypes.number.isRequired,
  onChangePage: PropTypes.func.isRequired,
  page: PropTypes.number.isRequired,
  rowsPerPage: PropTypes.number.isRequired,
};

InternalTablePaginationActions.defaultProps = {
  classes: {},
};

export default withStyles(styles)(InternalTablePaginationActions);
