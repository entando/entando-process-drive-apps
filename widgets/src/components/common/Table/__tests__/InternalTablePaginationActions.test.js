import React from 'react';
import { cleanup, render } from '@testing-library/react';

import InternalTablePaginationActions from 'components/common/Table/InternalTablePaginationActions';

describe('<InternalTablePaginationActions />', () => {
  afterEach(cleanup);

  it('renders snapshot correctly', () => {
    const { container } = render(
      <InternalTablePaginationActions
        page={0}
        count={42}
        rowsPerPage={5}
        onChangePage={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('check for pagination buttons when page === 0', () => {
    const { container } = render(
      <InternalTablePaginationActions
        page={0}
        count={42}
        rowsPerPage={5}
        onChangePage={jest.fn()}
      />
    );

    const buttons = container.querySelectorAll('button');
    expect(buttons[0].hasAttribute('disabled')).toBe(true);
    expect(buttons[1].hasAttribute('disabled')).toBe(true);
    expect(buttons[2].hasAttribute('disabled')).toBe(false);
    expect(buttons[3].hasAttribute('disabled')).toBe(false);
  });

  it('check for pagination buttons when page === 2', () => {
    const { container } = render(
      <InternalTablePaginationActions
        page={2}
        count={42}
        rowsPerPage={5}
        onChangePage={jest.fn()}
      />
    );

    const buttons = container.querySelectorAll('button');
    expect(buttons[0].hasAttribute('disabled')).toBe(false);
    expect(buttons[1].hasAttribute('disabled')).toBe(false);
    expect(buttons[2].hasAttribute('disabled')).toBe(false);
    expect(buttons[3].hasAttribute('disabled')).toBe(false);
  });

  it('check for pagination buttons when page is last page', () => {
    const { container } = render(
      <InternalTablePaginationActions
        page={5}
        count={42}
        rowsPerPage={10}
        onChangePage={jest.fn()}
      />
    );

    const buttons = container.querySelectorAll('button');
    expect(buttons[0].hasAttribute('disabled')).toBe(false);
    expect(buttons[1].hasAttribute('disabled')).toBe(false);
    expect(buttons[2].hasAttribute('disabled')).toBe(true);
    expect(buttons[3].hasAttribute('disabled')).toBe(true);
  });
});
