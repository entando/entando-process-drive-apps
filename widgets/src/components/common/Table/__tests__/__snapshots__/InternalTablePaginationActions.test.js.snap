// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InternalTablePaginationActions /> renders snapshot correctly 1`] = `
<div>
  <div
    class="InternalTablePaginationActions-root-1"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
          />
        </svg>
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root InternalTablePaginationActions-previousIcon-2"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
      </span>
    </button>
    <div
      class="MuiFormControl-root MuiTextField-root InternalTablePaginationActions-pageNumberInput-3"
      style="width: 35px;"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input"
          type="text"
          value="1"
        />
      </div>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1"
    >
      undefined 9
    </p>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
